### PR TITLE
core: enable binary compatability check

### LINKED ISSUE #9500: Enable binary compatibility check for 4.0.x branch of Micronaut Core
I have not checked but you may need to set a baseline. 

```groovy
micronautBuild {
    binaryCompatibility {
        baselineVerison = “4.0.0-RC3”
        enabled = true 
    }
}
```

### PR DESCRIPTION
**Pull request by wetted**
closes #9500

### CODE REVIEW

**Comment by wetted** [comment]
see https://github.com/micronaut-projects/micronaut-build/issues/584

**Review by sdelamo — CHANGES_REQUESTED** [review]
binary compatibility is still disabled. See https://github.com/micronaut-projects/micronaut-core/blob/4.0.x/buildSrc/src/main/groovy/io.micronaut.build.internal.convention-library.gradle#L20 or https://github.com/micronaut-projects/micronaut-core/blob/4.0.x/context-propagation/build.gradle#L9

File: core-bom/build.gradle [hunk 1]
```diff
@@ -15,9 +15,33 @@ micronautBom {
 
 micronautBuild {
     binaryCompatibility {
-        def dash = version.indexOf('-')
-        def v = dash > 0 ? version.substring(0, dash) : version
-        def (major, minor, patch) = v.split('[.]').collect { it.toInteger() }
-        enabled = major > 4 || (major == 4 && minor > 0) || (major == 4 && minor == 0 && patch > 0)
+        baselineVersion = "4.0.0-RC3"
```
Comment (sdelamo) [inline core-bom/build.gradle:18]: doe we need to set the baseline version?

Comment (wetted) [reply core-bom/build.gradle:18]: Apparently, so. If I try to build it resolves 3.9.4 and that causes problems with non-existent modules and `japiCmp` checks. @melix said we will need to remove it after the 4.0.0 release

**Comment by wetted** [comment]
I am having some pain getting this to work. I wonder if we need a new build plugin with the fix for `binaryCompatibility`.  @melix ?

File: core-bom/build.gradle (hunk 1 above)
Comment (wetted) [reply core-bom/build.gradle:18]: https://github.com/micronaut-projects/micronaut-core/issues/9500#issuecomment-1609603337

**Comment by melix** [comment]
let me check. I have pushed a fix on `micronaut-build` but I'll double check nothing else is needed

**Comment by melix** [comment]
There are more changes needed. In addition it seems that `micronaut-core` supplies the wrong group for the baseline comparator... On it.

**Comment by sdelamo** [comment]
@wetted @melix any idea about the sonar failure which made the build fail?

**Comment by melix** [comment]
Most likely due to the sonar plugin upgrade in the latest release of build plugins. Apparently the plugin expects a String somewhere but gets a GString. Trying to figure out where...

**Comment by melix** [comment]
See https://github.com/micronaut-projects/micronaut-build/pull/587

**Pull request merged**